### ISSUE TITLE
Grain pipeline does not reshuffle across epochs

### ISSUE THREAD

**Issue by nikhilr612** [nikhilr612/tinyflow#13]
## Summary

The Grain pipeline emits the **identical batch order and identical augmentation draws every
epoch**, so there is no shuffling across epochs and the augmentation is a one-shot fixed
perturbation of the dataset.

## Reproduction

```python
arr = np.load(".preprocessed/anime_faces.npy", mmap_mode="r")[:512]
ds = wrap_dataset(np.ascontiguousarray(arr), batch_size=32, seed=49)
for epoch in range(3):
    first = next(iter(ds))
    print(epoch, float(first.sum()), float(first[0, 0, 0, 0]))
```

Output:

```
epoch 0: first-batch sum=121678.828125  px0=0.957935
epoch 1: first-batch sum=121678.828125  px0=0.957935
epoch 2: first-batch sum=121678.828125  px0=0.957935
```

Bit-identical across epochs.

## Cause

`grain.MapDataset.shuffle(seed)` and `.random_map(...)` derive their randomness from the
**element index**, not from an iteration counter. `train_on_image` iterates the finite
`IterDataset` once per epoch, so indices run `0..N-1` every time and every epoch sees the same
composition, the same horizontal flips and the same colour jitter.

## Consequences

- No cross-epoch shuffling, so batch composition is a fixed partition for the whole run.
- `RandomHorizontalFlip(p=0.5)` effectively becomes "half the dataset is permanently mirrored"
  rather than an augmentation.
- Likewise `ColorJitter` applies one fixed perturbation per image for the entire run.

## Proposed fix

Insert `.repeat(n_epochs)` before the random maps so indices run `0..N*E-1` and each pass gets
fresh randomness, then restructure the training loop as a single step loop:

```python
for step, batch in enumerate(dataset, start=1):
    ...
    epoch = step // steps_per_epoch
```

which is simpler than the current nested loop and makes per-step metric records natural.

Minor, while here: `shuffle_seed = prng.getrandbits(7)` (`data/animefaces.py:159`) draws a
7-bit shuffle seed — 128 distinct orderings. Pass `seed` through instead.

**Comment by nikhilr612** [comment]
Fixed: `data/animefaces.py` (wrap_dataset, ~line 289) now derives per-call `shuffle_seed`/`augment_seed` via `random.Random(seed)`, and `shuffle` is documented as epoch-aware, so batch order and augmentation draws vary across epochs (commit 7ad2eed / a074a97). Confirmed in current code.
